File: solutions/cs_solutions/p3.cs
public class p3
{
    // Longest Substring Without Repeating Characters - Medium

    // Given a string, find the length of the longest substring without repeating characters.
    public class Solution
    {
        public int LengthOfLongestSubstring(string s)
        {
            if (string.IsNullOrEmpty(s)) return 0;
            if (s.Trim().Length == 0) return 1;

            var map = new Dictionary<char, int>();
            int maxLen = 0;
            int f = 0;
            for (int l = 0; l < s.Length; l++)
            {
                var c = s[l];
                if (map.ContainsKey(c))
                {
                    f = Math.Max(map[c], f);
                }
                if (maxLen < l - f + 1) maxLen = l - f + 1;
                map[c] = l + 1;
            }
            return maxLen;
        }
    }
}

File: solutions/cpp_solutions/p3.cpp
#include <string>
#include <unordered_map>
#include <algorithm>
using namespace std;

// Longest Substring Without Repeating Characters - Medium
// Given a string, find the length of the longest substring without repeating characters.

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> set{};
        int maxLength{0};
        int start{0};
        int i{0};
        for (; i != s.size(); ++i){
            unordered_map<char,int>::const_iterator got = set.find(s.at(i));
            if (got != set.end() && got->second >= start)
                start = got->second + 1;
            set[s.at(i)] = i;
            maxLength = max(maxLength, i - start + 1);
        }
        maxLength = max(maxLength, i - start);
        return maxLength;
    }
};